### PR TITLE
Add POST property endpoint for catalogue categories #259

### LINKED ISSUE #259: Add POST property endpoint for adding a single property to a catalogue category and its children
From discussions:

- [x] Add endpoint `/v1/catalogue-categories/{catalogue_category_id}/properties`
- [x] Allow POST on this endpoint with a catalogue item property and a `default_value` parameter to populate the existing values and it should then
   - [x] Add the property to the catalogue category
   - [x] Add the property to any child catalogue items
   - [x] Add the property to any child items of those catalogue items
   - [x] Reject the request if new property is mandatory and `default_value` is None
   - [x] `modified_time` should be modified at least at the catalogue category level, but if possible everywhere

We can likely use an `update_many` with a `$push` for updating items and maybe the catalogue items as well but we will likely need to iterate though the catalogue items ourselves.

We need a new service either `Property` or `PropertyMigration` to perform this operation using a transaction to ensure the whole thing either succeeds or fails as a whole. We should use the existing repos but with additional methods to do the `update_many`.

The `default_value` should be used to populate fields when adding them with a default value of None, so that optional fields store None by default, and the service layer should raise an error to reject the request if the property is mandatory but the `default_value` is None.

### PR DESCRIPTION
**Pull request by joelvdavies**
## Description
See #245.

This PR should be working with https://github.com/ral-facilities/inventory-management-system/pull/530.

## Notes
- Renames `CatalogueItemPropertyPostRequestSchema` -> `CatalogueCategoryPostRequestPropertySchema` so the new endpoint can use `CatalogueItemPropertyPostRequestSchema` as it requires a slightly different schema. This new name also reflects its location slightly better, although it technically should be `CatalogueCategoryPostRequestCatalogueItemPropertySchema` under the existing naming scheme, but that is very long. Further changes e.g. to `catalogue_item_properties` are left to a future issue
- Should we also change the type of property values to `Union[str, int, bool]` rather than Any?
- As we need the ids of catalogue items to know which items to update we can either use update_many with the catalogue_category_id or find the ids for the catalogue items first and then use $in - performance testing this with 100000 catalogue items showed that the first method took 0.97s, and the 2nd 1.57s. So I went with the former.
- ~For the items update, we could do an iteration in python calling update_many with each catalogue_item_id - a rough test instead using catalogue items (which is not an accurate example due to the test effectively being 100000 update_ones's) made it take 56.1 seconds rather than 1.57 seconds achieved when using $in. Bulk write updates are also another option but it sounds like they should be avoided unless the update data changes in some way. Testing it took 4.60 seconds. So it appears overall the fastest method for items is using `$in` with the ids. The size of the array for this appears to be around 27 MB (using https://code.activestate.com/recipes/577504/), so over time we should be doing this update in batches (although current get requests for such a category would have a much higher memory usage)~. **UPDATE**: Now use distinct, which decreases the memory usage significantly (previous tests stored a dictionary of _id: ObjectId)
- Using multiple workers allows multiple requests to occur at once - adding a time.sleep inside the transaction and then editing on the front end will wait until the transaction is complete before completing the edit request if it is inside of the modified catalogue category. If it is not then the request happens instantly as long as there is another worker available on the api. If the sleep is before any edits within the transaction, the request is handled instantly provided there is a worker.
- Used a class in the e2e tests as they were going to be very nasty without (the unit tests really need something similar as we have too many arguments and locals) obviously may change later anyway but I think it could work nicely with the new test client as well

## Testing instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage

## Agile board tracking
Closes #259

### CODE REVIEW

File: inventory_management_system_api/schemas/catalogue_category.py [hunk 1]
```diff
@@ -174,3 +174,40 @@ class CatalogueCategorySchema(CreatedModifiedSchemaMixin, CatalogueCategoryPostR
...
+        :param allowed_values: The value of the `allowed_values` field.
+        :param info: Validation info from pydantic.
+        :return: The value of the `allowed_values` field.
+        """
+        if default_value is not None:
+            if not CatalogueCategoryPostRequestPropertySchema.is_valid_property_type(
+                expected_property_type=info.data["type"], property_value=default_value
+            ):
+                raise ValueError("default_value must be the same type as the property itself")
+            if info.data["allowed_values"] and default_value not in info.data["allowed_values"]:
```
Comment (joshuadkitenge) [inline inventory_management_system_api/schemas/catalogue_category.py:210]: This need to check the allowed_values.values and no the allowed_values object

Comment (joshuadkitenge) [inline inventory_management_system_api/schemas/catalogue_category.py:210]:
```suggestion
            if info.data["allowed_values"] and default_value not in info.data["allowed_values"].values:
```

File: test/unit/repositories/test_catalogue_category.py [hunk 2]
```diff
@@ -1145,3 +1146,43 @@ def test_has_child_elements_with_child_catalogue_items(test_helpers, database_mo
     result = catalogue_category_repository.has_child_elements(catalogue_category_id)
 
     assert result
+
+
+@patch("inventory_management_system_api.repositories.catalogue_category.datetime")
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply test/unit/repositories/test_catalogue_category.py:1151]: I don't think there is necessarily any need to in this test, its actual value is irrelevant, its really just checking its using the value that datetime_mock would have returned, which if not specified is a MagicMock by default.

File: test/unit/repositories/test_item.py [hunk 3]
```diff
@@ -588,3 +590,31 @@ def test_update_with_invalid_id(item_repository):
...
+    """
+    Test inserting a catalogue item property
+
+    Verify that the `insert_property_to_all_matching` method properly handles the insertion of a
+    property
+    """
+    session = MagicMock()
+    catalogue_item_ids = [ObjectId(), ObjectId()]
+    property_in = PropertyIn(**MOCK_PROPERTY_A_INFO)
+    datetime_mock.now.return_value = datetime(2024, 2, 16, 14, 0, 0, 0, tzinfo=timezone.utc)
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply test/unit/repositories/test_item.py:606]: 😄  good spot, in fact like above I don't think I even needed to set it

**Comment by VKTB** [comment]
When sending the request below
```
curl -X 'POST' \
  'http://localhost:8000/v1/catalogue-categories/65d8616ec22eeec8a0913249/properties' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "name": "Property C",
  "type": "boolean",
  "mandatory": false,
  "default_value": true
}'
```

The default value evaluates to 1 so the value for this property in the catalogue items and items gets set to 1
```json
[
  {
    "catalogue_category_id": "65d8616ec22eeec8a0913249",
    "properties": [
      {
        "id": "66431610ae5f84c1148c5e2b",
        "value": 1,
        "name": "Property C",
        "unit": null
      }
    ]
    ...
  }
]

**Comment by VKTB** [comment]
The following also works (probably related to the bug I reported in my previous comment)
```
curl -X 'POST' \
  'http://localhost:8000/v1/catalogue-categories/65d8616ec22eeec8a0913249/properties' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "name": "Property G",
  "type": "number",
  "unit": "cm",
  "mandatory": true,
  "allowed_values": {
      "type": "list",
      "values": [1, 3, 5, 7]
  },
  "default_value": true
}'
```

**Comment by joelvdavies** [comment]
> The following (where I supply a boolean default value for a number property) also works (probably related to the bug I reported in my previous comment)
> 
> ```
> curl -X 'POST' \
>   'http://localhost:8000/v1/catalogue-categories/65d8616ec22eeec8a0913249/properties' \
>   -H 'accept: application/json' \
>   -H 'Content-Type: application/json' \
>   -d '{
>   "name": "Property G",
>   "type": "number",
>   "unit": "cm",
>   "mandatory": true,
>   "allowed_values": {
>       "type": "list",
>       "values": [1, 3, 5, 7]
>   },
>   "default_value": true
> }'
> ```

Good spot, turns out boolean is a subclass of integer, so the type check needs to be more explicit using `type` instead of `isinstance` and ignoring the pylint warning `unidiomatic-typecheck`.

**Comment by VKTB** [comment]
For some reason, the below request fails with status code 500 but I was expecting to get a different status code with a message that either says unit is not allowed for a boolean property or allowed values are not allowed for a boolean property.
```
curl -X 'POST' \
  'http://localhost:8000/v1/catalogue-categories/65d8616ec22eeec8a0913249/properties' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "name": "Property G",
  "type": "boolean",
  "unit": "cm",
  "mandatory": true,
  "allowed_values": {
      "type": "list",
      "values": [1, 3, 5, 7]
  },
  "default_value": true
}'
```

Logs printed:
```
  File "/inventory-management-system-api/inventory_management_system_api/schemas/catalogue_category.py", line 214, in validate_default_value
    if info.data["allowed_values"] and default_value not in info.data["allowed_values"].values:
       ~~~~~~~~~^^^^^^^^^^^^^^^^^^
KeyError: 'allowed_values'
INFO:    127.0.0.1:62825 - "POST /v1/catalogue-categories/65d8616ec22eeec8a0913249/properties HTTP/1.1" 500
```

**Comment by joelvdavies** [comment]
> For some reason, the below request fails with status code 500 but I was expecting to get a different status code with a message that either says unit is not allowed for a boolean property or allowed values are not allowed for a boolean property.
> 
> ```
> curl -X 'POST' \
>   'http://localhost:8000/v1/catalogue-categories/65d8616ec22eeec8a0913249/properties' \
>   -H 'accept: application/json' \
>   -H 'Content-Type: application/json' \
>   -d '{
>   "name": "Property G",
>   "type": "boolean",
>   "unit": "cm",
>   "mandatory": true,
>   "allowed_values": {
>       "type": "list",
>       "values": [1, 3, 5, 7]
>   },
>   "default_value": true
> }'
> ```
> 
> Logs printed:
> 
> ```
>   File "/inventory-management-system-api/inventory_management_system_api/schemas/catalogue_category.py", line 214, in validate_default_value
>     if info.data["allowed_values"] and default_value not in info.data["allowed_values"].values:
>        ~~~~~~~~~^^^^^^^^^^^^^^^^^^
> KeyError: 'allowed_values'
> INFO:    127.0.0.1:62825 - "POST /v1/catalogue-categories/65d8616ec22eeec8a0913249/properties HTTP/1.1" 500
> ```

Just fixed that as well in the latest commit, I noticed it too 😄

**Comment by joelvdavies** [comment]
> For some reason, the below request fails with status code 500 but I was expecting to get a different status code with a message that either says unit is not allowed for a boolean property or allowed values are not allowed for a boolean property.
> 
> ```
> curl -X 'POST' \
>   'http://localhost:8000/v1/catalogue-categories/65d8616ec22eeec8a0913249/properties' \
>   -H 'accept: application/json' \
>   -H 'Content-Type: application/json' \
>   -d '{
>   "name": "Property G",
>   "type": "boolean",
>   "unit": "cm",
>   "mandatory": true,
>   "allowed_values": {
>       "type": "list",
>       "values": [1, 3, 5, 7]
>   },
>   "default_value": true
> }'
> ```
> 
> Logs printed:
> 
> ```
>   File "/inventory-management-system-api/inventory_management_system_api/schemas/catalogue_category.py", line 214, in validate_default_value
>     if info.data["allowed_values"] and default_value not in info.data["allowed_values"].values:
>        ~~~~~~~~~^^^^^^^^^^^^^^^^^^
> KeyError: 'allowed_values'
> INFO:    127.0.0.1:62825 - "POST /v1/catalogue-categories/65d8616ec22eeec8a0913249/properties HTTP/1.1" 500
> ```

This should be fixed as well, adding another e2e test as it wasn't covered.

**Comment by VKTB** [comment]
Sending the request below still sets the value for a boolean property to a number
```
curl -X 'POST' \
  'http://localhost:8000/v1/catalogue-categories/65d8616ec22eeec8a0913249/properties' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "name": "Property H",
  "type": "boolean",
  "mandatory": true,
  "default_value": false
}'
```

```json
[
  {
    "catalogue_category_id": "65d8616ec22eeec8a0913249",
    "properties": [
      {
        "id": "66432bd4fff58b6e9dd907b9",
        "value": 0,
        "name": "Property H",
        "unit": null
      }
    ]
    ...
  },
]

**Review by VKTB — APPROVED** [review]

**Pull request merged**